### PR TITLE
[irods/irods#5311] Overhaul help2man manpage generation (master)

### PR DESCRIPTION
**Pull request by SwooshyCueb**
Addresses irods/irods#5311 for master.

Notable changes:
- Manpage generation can be toggled with the CMake variable `BUILD_DOCS`. This defaults to `ON` if suitable `help2man` and `gzip` tools are found.
- `help2man` and `gzip` are now invoked directly from the buildsystem with no helper shell script. This allows us to leverage CMake's `find_program` functionality to allow the developer to manually specify the path to `help2man` and `gzip`.
- New `docs` target. This influences build order to group manpage generation targets together when building `all`.
- The directory containing `libirods_client.so` is prepended to `LD_LIBRARY_PATH` when invoking `help2man`. This solves/circumvents the transient dependency issue.

### CODE REVIEW

**Comment by SwooshyCueb** [comment]
Pinging @trel in case you haven't seen it

File: CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,76 @@ if (NOT CMAKE_CONFIGURATION_TYPES AND NOT CMAKE_BUILD_TYPE)
...
+      execute_process(
+        COMMAND "${HELP2MAN}" --version
+        OUTPUT_VARIABLE HELP2MAN_VERSION
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        )
+      string(REGEX REPLACE "\n.*" "" HELP2MAN_VERSION "${HELP2MAN_VERSION}")
+      string(REGEX REPLACE "[^ ]+ +[^ ]+ +" "" HELP2MAN_VERSION "${HELP2MAN_VERSION}")
+      message(STATUS "Found help2man version ${HELP2MAN_VERSION}")
+      if (HELP2MAN_VERSION VERSION_LESS "1.37")
+        # don't cache unsable help2man
```
Comment (trel) [inline CMakeLists.txt:49]: unsable -> unusable

File: CMakeLists.txt [hunk 2]
```diff
@@ -21,6 +21,76 @@ if (NOT CMAKE_CONFIGURATION_TYPES AND NOT CMAKE_BUILD_TYPE)
...
+endif()
+
+# Find gzip
+if (NOT DEFINED BUILD_DOCS OR BUILD_DOCS)
+  if (NOT DEFINED GZIP)
+    set(GZIP_SET_BY_CMAKE ON)
+  endif()
+  find_program(
+    GZIP
+    NAMES gzip pigz
```
Comment (trel) [inline CMakeLists.txt:74]: i didn't know about pigz!

File: CMakeLists.txt [hunk 3]
```diff
@@ -169,17 +251,28 @@ foreach(IRODS_CLIENT_ICOMMANDS_SCRIPT ${IRODS_CLIENT_ICOMMANDS_SCRIPTS})
...
-    ${CMAKE_BINARY_DIR}/man/${IRODS_CLIENT_ICOMMANDS_SCRIPT}.1.gz
-    DESTINATION ${CMAKE_INSTALL_MANDIR}/man1
-    )
+  if (BUILD_DOCS)
+    set(SCRIPT_MANPAGE "${CMAKE_BINARY_DIR}/man/${IRODS_CLIENT_ICOMMANDS_SCRIPT}.1")
+    add_custom_command(
+      OUTPUT ${SCRIPT_MANPAGE}
+      DEPENDS ${SCRIPT_SRC_PATH}
+      COMMENT "generating manpage ${IRODS_CLIENT_ICOMMANDS_SCRIPT}.1"
+      COMMAND "${CMAKE_COMMAND}" -E env "LD_LIBRARY_PATH=$<TARGET_FILE_DIR:irods_client>:$ENV{LD_LIBRARY_PATH}" "${HELP2MAN}" -h -h -N -n "an iRODS iCommand" --version-string="iRODS-${IRODS_VERSION}" "${SCRIPT_SRC_PATH}" -o "${SCRIPT_MANPAGE}"
```
Comment (trel) [inline CMakeLists.txt:260]: surprised to see the help2man invocation defined twice...

i have no suggestions, just noticing slight duplication.

**Review by trel — COMMENTED** [review]
one more shell script gone...

**Comment by trel** [comment]
please also note/incorporate/consider PR https://github.com/irods/irods_client_icommands/pull/121

File: CMakeLists.txt (hunk 2 above)
Comment (SwooshyCueb) [reply CMakeLists.txt:74]: I wanted to stick `bgzip` in there, too, but it doesn't support all the `gzip` arguments we use.

File: CMakeLists.txt (hunk 3 above)
Comment (SwooshyCueb) [reply CMakeLists.txt:260]: I did consider implementing a macro and sticking it, and all the `find_program` stuff in a cmake module, but I decided that I might do it as part of work towards 5250.

Comment (trel) [reply CMakeLists.txt:260]: seems fair.

**Comment by trel** [comment]
happy to merge his, so credit is retained, and then make any remaining changes here yourself.

**Comment by trel** [comment]
anything else left to do here?

does it play well with #121?   does it matter?

**Comment by SwooshyCueb** [comment]
Turns out, there's more than just the iquest examples that `help2man` bungles up. I've done my best to format the help output of iquest to look good in both locations, but honestly I'm starting to think we might want to consider other approaches. There are several other manpages that need attention, and it would be easier to write them by hand than it would be to wrangle `help2man`.

**Comment by trel** [comment]
definitely - the iquest PR is/was a WIP.

a very reasonable position on the 'other approaches' - `help2man` was a thing to not have to touch 50 iCommands the first time we needed to compile all these for the web.

open to not having man pages at all - but that's a bit hard to justify - so i think we're looking at touching the help text a bit more unless we come up with an alternate plan.

**Comment by SwooshyCueb** [comment]
As discussed, I've opened an issue for the help2man formatting issues so that we can go ahead and merge this.

\#'d

**Pull request merged**